File: Lib/Java/com.scbeta.helpers/src/com/scbeta/helpers/XmlTreeNode.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package com.scbeta.helpers;

import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;
import org.dom4j.Attribute;
import org.dom4j.Document;
import org.dom4j.DocumentException;
import org.dom4j.DocumentHelper;
import org.dom4j.Element;

/**
 *
 * @author aaa
 */
public class XmlTreeNode {

    private String key;
    private String value;
    private List<XmlTreeNode> items;
    private Map<String, String> attribs;
    private static TextEncoder xmlTextEncoder = TextEncoder.getXmlTextEncoder();
    private static TextEncoder jsonTextEncoder = TextEncoder.getJsonTextEncoder();
    // 获取结点名称

    public String getKey() {
        return key;
    }

    public String getEncodedKey() {
        return xmlTextEncoder.encodeText(key);
    }

    //设置结点名称
    public void setKey(String value) {
        key = value;
    }

    // 获取结点的值
    public String getValue() {
        if (value == null) {
            return "";
        }
        return value;
    }

    public String getEncodedValue() {
        if (value == null) {
            return "";
        }
        return xmlTextEncoder.encodeText(value);
    }

    // 设置结点的值(Value与Items同时只能有一个有效)
    public void setValue(String value) {
        if (value == null) {
            this.value = "";
        } else {
            this.value = xmlTextEncoder.decodeText(value);
        }
    }

    // 子结点集合(Value与Items同时只能有一个有效)
    public List<XmlTreeNode> getItems() {
        return items;
    }

    // 属性集合
    public Map<String, String> getAttribs() {
        return attribs;
    }

    // 构造函数
    public XmlTreeNode() {
        init("", "");
    }

    // 构造函数
    public XmlTreeNode(String key) {
        init(key, "");
    }

    // 构造函数
    public XmlTreeNode(String key, String value) {
        init(key, value);
    }

    // 初始化
    private void init(String key, String value) {
        this.setKey(key);
        this.setValue(value);

        items = new LinkedList<XmlTreeNode>();
        attribs = new LinkedHashMap<String, String>();
    }

    // 根据Key得到属性
    public String getAttributeValue(String key) {
        if (attribs.containsKey(key)) {
            return attribs.get(key);
        } else {
            return null;
        }
    }

    // 设置属性
    public void setAttribute(String key, String value) {
        if (attribs.containsKey(key)) {
            attribs.remove(key);
        }
        attribs.put(key, value);
    }

    //移除属性
    public void removeAttribute(String key) {
        attribs.remove(key);
    }

    public void removeAllAttribute() {
        attribs.clear();
    }

    // 根据Key得到对象
    private XmlTreeNode getChildItem(String key) {
        for (XmlTreeNode item : items) {
            if (item.getKey().equals(key)) {
                return item;
            }
        }
        return null;
    }

    // 根据Key得到其子对象列表
    private List<XmlTreeNode> getChildItems(String key) {
        List<XmlTreeNode> lstNodes = new LinkedList<XmlTreeNode>();
        for (XmlTreeNode item : items) {
            if (item.getKey().equals(key)) {
                lstNodes.add(item);
            }
        }
        return lstNodes;
    }

    // 根据路径得到对象
    public XmlTreeNode getItem(String path) {
        if (path == null || path.isEmpty()) {
            return this;
        }
        XmlTreeNode TmpTn = this;

        String[] TmpKeys = path.split("/");
        for (String tmpKey : TmpKeys) {
            String tmpKeyStr = tmpKey.trim();
            if (tmpKeyStr == null || tmpKeyStr.equals("")) {
                continue;
            }

            TmpTn = TmpTn.getChildItem(tmpKeyStr);
            if (TmpTn == null) {
                return null;
            }
        }
        return TmpTn;
    }

    //移除指定的对象
    public void removeItem(XmlTreeNode node) {
        items.remove(node);
    }

    //移除所有的对象
    public void removeAllItems() {
        items.clear();
    }

    /// <summary>
    /// 根据路径得到对象集合
    /// </summary>
    /// <param name="Path">路径</param>
    /// <returns></returns>
    public List<XmlTreeNode> getItems(String path) {
        XmlTreeNode ParentNode = this;
        String[] TmpKeys = path.split("/");
        String LastKey = TmpKeys[TmpKeys.length - 1];
        if (TmpKeys.length > 1) {
            String ParentPath = path.substring(0, path.length() - (LastKey.length() + 1));
            ParentNode = getItem(ParentPath);
            if (ParentNode == null) {
                return new LinkedList<XmlTreeNode>();
            }
        }
        return ParentNode.getChildItems(LastKey);
    }

    /// <summary>
    /// 根据路径得到对象的值
    /// </summary>
    /// <param name="Path">路径</param>
    /// <returns></returns>
    public String getItemValue(String path) {
        XmlTreeNode TmpTn = getItem(path);
        if (TmpTn == null) {
            return null;
        }
        return TmpTn.getValue();
    }

    // 添加子结点
    public XmlTreeNode addItem(XmlTreeNode tn) {
        items.add(tn);
        return tn;
    }

    // 添加子结点
    public XmlTreeNode addItem(String key) {
        return addItem(key, "");
    }

    // 添加子结点
    public XmlTreeNode addItem(String key, String value) {
        XmlTreeNode tn = new XmlTreeNode(key);
        tn.setValue(value);
        return addItem(tn);
    }

    //添加子结点
    public XmlTreeNode addItem(String key, XmlTreeNode node) {
        XmlTreeNode tn = new XmlTreeNode(key);
        tn.addItem(node);
        addItem(tn);
        return tn;
    }

    // ToString方法(此方法仅供调试时查看对象的值，要生成XML请用toXml方法)
    @Override
    public String toString() {
        return this.toXml();
    }

    //输出XML
    public String toXml() {
        return XmlTreeNode.toXml(this, "utf-8");
    }

    public String toXml(String encoding) {
        return XmlTreeNode.toXml(this, encoding);
    }

    //输出Json
    public String toJson() {
        return XmlTreeNode.toJson(this);
    }

    // 输出XML
    public static String toXml(XmlTreeNode root, String encoding) {
        String xml = "";

        org.dom4j.DocumentFactory factory = org.dom4j.DocumentFactory.getInstance();
        Document document = factory.createDocument();
        //根结点
        Element rootElement = document.addElement(root.getKey());
        //设置XML文档编码
        document.setXMLEncoding(encoding);

        _toXml(root, rootElement);
        xml = document.asXML();
        return xml;
    }

    private static void _toXml(XmlTreeNode treeNode, Element parentXmlElement) {
        //先添加属性
        for (String key : treeNode.getAttribs().keySet()) {
            parentXmlElement.addAttribute(key, treeNode.getAttribs().get(key));
        }

        //再添加值或子结点
        if (treeNode.getItems().isEmpty()) {
            String tmpValue = treeNode.getEncodedValue();
            if (tmpValue == null) {
                tmpValue = "";
            }
            parentXmlElement.setText(tmpValue);
        } else {
            for (XmlTreeNode item : treeNode.getItems()) {
                Element newXmlElement = parentXmlElement.addElement(item.getKey());
                _toXml(item, newXmlElement);
            }
        }
    }

    // 从XML得到TreeNode对象
    public static XmlTreeNode fromXml(String xml) {
        if (xml == null || xml.isEmpty()) {
            return null;
        }
        Document document = null;
        try {
            document = DocumentHelper.parseText(xml);
        } catch (DocumentException ex) {
            ConsoleHelper.println("fromXml:" + xml);
            Logger.getLogger(Logger.GLOBAL_LOGGER_NAME).log(Level.SEVERE, xml, ex);
            return null;
        }

        Element XmlRoot = document.getRootElement();
        return _fromXml(XmlRoot);
    }

    private static XmlTreeNode _fromXml(Element parentXmlElement) {
        XmlTreeNode TnNode = new XmlTreeNode(parentXmlElement.getName());

        //先写属性
        for (Attribute attrib : (List<Attribute>) parentXmlElement.attributes()) {
            TnNode.setAttribute(attrib.getName(), attrib.getValue());
        }

        //再写子结点
        if (parentXmlElement.elements().isEmpty()) {
            String tmpValue = parentXmlElement.getTextTrim();
            if (tmpValue == null) {
                tmpValue = "";
            }
            TnNode.setValue(tmpValue);
        } else {
            for (Element childElement : (List<Element>) parentXmlElement.elements()) {
                TnNode.addItem(_fromXml(childElement));
            }
        }
        return TnNode;
    }

    public static String toJson(XmlTreeNode treeNode) {
        StringBuilder sb = new StringBuilder();
        _toJson(treeNode, sb);
        return sb.toString();
    }

    private static void _toJson(XmlTreeNode treeNode, StringBuilder sb) {
        //如果当前是根结点
        boolean isRootNode = sb.length() == 0;
        //如果当前根结点的名字是root
        boolean isRootNodeName_root = "root".equals(treeNode.getKey());

        sb.append("{");
        if (isRootNode && !isRootNodeName_root) {
            sb.append(String.format("\"%s\":{", treeNode.getKey()));
        }
        //添加属性
        boolean isFirstAttrib = true;
        for (String attribKey : treeNode.getAttribs().keySet()) {
            if (!isFirstAttrib) {
                sb.append(",");
            }
            isFirstAttrib = false;

            sb.append(String.format("\"%s\":\"%s\"", attribKey, jsonTextEncoder.encodeText(treeNode.getAttributeValue(attribKey))));
        }
        //添加子结点
        boolean isFirstNode = true;
        for (XmlTreeNode childNode : treeNode.getItems()) {
            if (treeNode.getAttribs().keySet().size() > 0 || !isFirstNode) {
                sb.append(",");
            }
            isFirstNode = false;

            if (childNode.getKey() != null || !childNode.getKey().isEmpty()) {
                sb.append(String.format("\"%s\":", childNode.getKey()));
            }
            _toJson(childNode, sb);
        }
        //添加大括号结束
        if (isRootNode && !isRootNodeName_root) {
            sb.append("}");
        }
        sb.append("}");
    }

    public static XmlTreeNode fromJson(String json) throws Exception {
        throw new Exception("未实现的方法！此方法还未实现");
    }
}

File: trunk/Lib/DotNet/aaaSoft.Helpers/XmlTreeNode.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using System.Xml;

namespace aaaSoft.Helpers
{
    public class XmlTreeNode
    {
        private String _key;
        private String value;
        private List<XmlTreeNode> items;
        private Dictionary<String, String> attribs;


        /// <summary>
        /// 获取或设置结点名称
        /// </summary>
        public String Key
        {
            get { return _key; }
            set { this._key = value; }
        }
                
        /// <summary>
        /// 获取或设置结点的值(Value与Items同时只能有一个有效)
        /// </summary>
        public String Value
        {
            get
            {
                if (String.IsNullOrEmpty(value))
                {
                    return String.Empty;
                }
                return value;
            }
            set
            {
            if (value == null)
            {
                this.value = String.Empty;
            }
            else
            {
                this.value = value;
            }
            }
        }        

        
        #region 构造函数
        /// <summary>
        /// 构造函数
        /// </summary>
        public XmlTreeNode()
            : this(String.Empty, String.Empty)
        { }

        /// <summary>
        /// 构造函数
        /// </summary>
        /// <param name="Key">结点名称</param>
        public XmlTreeNode(String Key)
            : this(Key, String.Empty)
        { }

        /// <summary>
        /// 构造函数
        /// </summary>
        /// <param name="key">结点名称</param>
        /// <param name="value">结点值</param>
        public XmlTreeNode(String key, String value)
        {
            this._key = key;
            this.value = value;
            items = new List<XmlTreeNode>();
            attribs = new Dictionary<string, string>();
        }
        #endregion

        #region 根据Key得到属性
        public String GetAttributeValue(String key)
        {
            foreach (var attrib in attribs)
            {
                if (attrib.Key == key)
                    return attrib.Value;
            }
            return null;
        }
        #endregion

        #region 添加属性
        /// <summary>
        /// 添加属性
        /// </summary>
        /// <param name="key"></param>
        /// <param name="value"></param>
        public void SetAttribute(String key, String value)
        {
            if (attribs.ContainsKey(key))
                attribs.Remove(key);
            attribs.Add(key, value);
        }
        #endregion

        //移除属性
        public void RemoveAttribute(String key)
        {
            attribs.Remove(key);
        }

        #region 根据Key得到对象
        /// <summary>
        /// 根据Key得到其子对象
        /// </summary>
        /// <param name="key">名称</param>
        /// <returns></returns>
        private XmlTreeNode GetChildItem(String key)
        {
            foreach (var item in items)
            {
                if (item._key == key)
                {
                    return item;
                }
            }
            return null;
        }

        /// <summary>
        /// 根据Key得到其子对象列表
        /// </summary>
        /// <param name="key">名称</param>
        /// <returns></returns>
        private List<XmlTreeNode> GetChildItems(String key)
        {
            List<XmlTreeNode> lstNodes = new List<XmlTreeNode>();
            foreach (var item in items)
            {
                if (item._key == key)
                {
                    lstNodes.Add(item);
                }
            }
            return lstNodes;
        }

        /// <summary>
        /// 根据路径得到对象
        /// </summary>
        /// <param name="Path">路径</param>
        /// <returns></returns>
        public XmlTreeNode GetItem(String Path)
        {
            if (String.IsNullOrEmpty(Path))
                return this;

            XmlTreeNode TmpTn = this;

            var TmpKeys = Path.Split('/');
            foreach (var tmpKey in TmpKeys)
            {
                String tmpKeyStr = tmpKey.Trim();
                if (String.IsNullOrEmpty(tmpKeyStr)) continue;

                TmpTn = TmpTn.GetChildItem(tmpKeyStr);
                if (TmpTn == null) return null;
            }
            return TmpTn;
        }

        /// <summary>
        /// 得到对象集合
        /// </summary>
        /// <returns></returns>
        public List<XmlTreeNode> GetItems()
        {
            return this.items;
        }

        /// <summary>
        /// 根据路径得到对象集合
        /// </summary>
        /// <param name="Path">路径</param>
        /// <returns></returns>
        public List<XmlTreeNode> GetItems(String Path)
        {
            var ParentNode = this;
            var TmpKeys = Path.Split('/');
            var LastKey = TmpKeys[TmpKeys.Length - 1];
            if (TmpKeys.Length > 1)
            {
                var ParentPath = Path.Substring(0, Path.Length - (LastKey.Length + 1));
                ParentNode = GetItem(ParentPath);
                if (ParentNode == null)
                    return new List<XmlTreeNode>();
            }
            return ParentNode.GetChildItems(LastKey);
        }

        /// <summary>
        /// 根据路径得到对象的值
        /// </summary>
        /// <param name="path">路径</param>
        /// <returns></returns>
        public String GetItemValue(String path)
        {
            var TmpTn = GetItem(path);
            if (TmpTn == null) return null;
            return TmpTn.value;
        }
        #endregion

        #region 添加子结点
        /// <summary>
        /// 添加子结点
        /// </summary>
        /// <param name="tn">TreeNode对象</param>
        /// <returns>TreeNode对象</returns>
        public XmlTreeNode AddItem(XmlTreeNode tn)
        {
            items.Add(tn);
            return tn;
        }
        /// <summary>
        /// 添加子结点
        /// </summary>
        /// <param name="Key">结点名称</param>
        /// <returns>TreeNode对象</returns>
        public XmlTreeNode AddItem(String Key)
        {
            return AddItem(Key, String.Empty);
        }
        /// <summary>
        /// 添加子结点
        /// </summary>
        /// <param name="key">结点名称</param>
        /// <param name="value">结点值</param>
        /// <returns>TreeNode对象</returns>
        public XmlTreeNode AddItem(String key, String value)
        {
            XmlTreeNode tn = new XmlTreeNode(key);
            tn.value = value;
            AddItem(tn);
            return tn;
        }

        /// <summary>
        /// 添加子结点
        /// </summary>
        /// <param name="key">结点名称</param>
        /// <param name="node">结点</param>
        /// <returns></returns>
        public XmlTreeNode AddItem(String key, XmlTreeNode node)
        {
            XmlTreeNode tn = new XmlTreeNode(key);
            tn.AddItem(node);
            AddItem(tn);
            return tn;
        }
        #endregion


        #region ToString方法
        /// <summary>
        /// 此方法仅供调试时查看对象的值，要生成XML请用GenerateXml方法
        /// </summary>
        /// <returns></returns>
        public override string ToString()
        {
            return ToXml(this, Encoding.Default);
        }
        #endregion

        #region 生成XML
        /// <summary>
        /// 生成XML
        /// </summary>
        /// <param name="root">根结点</param>
        /// <returns></returns>
        public static String ToXml(XmlTreeNode root, Encoding encoding)
        {
            MemoryStream ms = new MemoryStream();
            XmlWriter xw = new System.Xml.XmlTextWriter(ms, encoding);

            xw.WriteStartDocument();
            _WriteTreeNode(root, xw);
            xw.WriteEndDocument();

            xw.Flush();
            xw.Close();

            var xml = encoding.GetString(ms.ToArray());
            return xml;
        }

        private static void _WriteTreeNode(XmlTreeNode tn, XmlWriter xw)
        {
            xw.WriteStartElement(tn._key);
            //先写属性
            foreach (var attrib in tn.attribs)
            {
                xw.WriteStartAttribute(attrib.Key);
                xw.WriteValue(attrib.Value);
                xw.WriteEndAttribute();
            }
            //再写值或子结点
            if (tn.items.Count == 0)
            {
                String tmpValue = tn.value;
                if (tmpValue == null) tmpValue = String.Empty;
                xw.WriteValue(tmpValue);
            }
            else
            {
                foreach (var item in tn.items)
                {
                    _WriteTreeNode(item, xw);
                }
            }
            xw.WriteEndElement();
        }
        #endregion

        #region 从XML得到TreeNode对象

        public String ToXml()
        {
            return XmlTreeNode.ToXml(this, new UTF8Encoding(false));
        }

        public String ToXml(Encoding encoding)
        {
            return XmlTreeNode.ToXml(this, encoding);
        }


        /// <summary>
        /// 从XML得到TreeNode对象
        /// </summary>
        /// <param name="xml">XML字符串</param>
        /// <returns></returns>
        public static XmlTreeNode FromXml(String xml)
        {
            XmlDocument document = new XmlDocument();
            try { document.LoadXml(xml); }
            catch { return null; }
            var XmlRoot = document.DocumentElement;
            return _fromXml(XmlRoot);
        }
        private static XmlTreeNode _fromXml(XmlNode XmlNode)
        {
            XmlTreeNode TnNode = new XmlTreeNode(XmlNode.Name);
            //先写属性
            foreach (XmlAttribute attrib in XmlNode.Attributes)
            {
                TnNode.SetAttribute(attrib.Name, attrib.Value);
            }
            //再写子结点
            if (XmlNode.ChildNodes.Count == 0)
            {
                String tmpValue = XmlNode.Value;
                if (tmpValue == null) tmpValue = String.Empty;
                TnNode.value = tmpValue;
            }
            else if (XmlNode.ChildNodes.Count == 1 && XmlNode.ChildNodes[0] is XmlText)
            {
                TnNode.value = XmlNode.ChildNodes[0].Value;
            }
            else
            {
                foreach (XmlNode childNode in XmlNode.ChildNodes)
                {
                    if (childNode is XmlElement)
                        TnNode.AddItem(_fromXml(childNode));
                }
            }
            return TnNode;
        }
        #endregion
    }
}
